index.d.ts

```diff
@@ -0,0 +1,18 @@
+interface ListGithubDirContentOptions {
+	user: string
+	repository: string
+	ref?: string
+	directory: string
+	token: string
+	getFullData?: boolean
+}
+
+function ListGithubDirContent<ProvidedOptions extends ListGithubDirContentOptions>(options: ProvidedOptions):
```

Comment:
I think this should work:

```suggestion
async function ListGithubDirContent(options: ListGithubDirContentOptions):
```

Reply:
This won't work. We need the generic to provide the correct return type based on the provided `getFullData` option.

Reply:
Are you sure about that? It seems to "compile" for me:

```ts

type ViaContentsApi = (options: ListGithubDirContentOptions) =>
ListGithubDirContentOptions['getFullData'] extends true ? Promise<any[]> : Promise<string[]>
type ViaTreesApi = (options: ListGithubDirContentOptions) =>
ListGithubDirContentOptions['getFullData'] extends true ? Promise<TreeResult<any>> : Promise<TreeResult<string>>
```

Reply:
It compiles, but it does not work. 😁

TypeScript will always provide `Promise<string[]>`/`Promise<TreeResult<string>>` as a return type, because `ListGithubDirContentOptions['getFullData']` is `boolean`, and therefore `ListGithubDirContentOptions['getFullData'] extends true` never passes. It's like writing `boolean extends true`.

Reply:
[See this](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgDLAM5gOLDACwFcAjAEWCgGEB7cCcAeQAcxhaNkBvAWACgBIQhmgAuZFiigA5n35QITahjzUoATzETps+TAD8msJJAyBAEwoQEYVRvFHtAmwGt6B+8dP8pEMADFCABtA0jgwOHdiampAiDgQPgBfPj5QSFhEFAAVeQgAJQgMILAAHiyAPmQIAA9IEDMOAEEoKDg1MsqeJyhCEAQwiDMxKJi4hN5k3j4wNSYUADVgOBo6cAxGpmBkAF5kEoAFKGoAN2AzQeZWdira+ga0TBw8IjIKFbqwS7YQDHKACmoLG+GDEhxOZwuQPYAEoduVZGDTuczF92ABtADkPn8QRCYTgGIAujc6vcjIQUHpkGCALaYCAleJqNGEyqgo504QlLQmFnwqa8GZzZCLZa0D7rTYAdWe1EIOHo0GACB2yABUJ+YnQWFwBBI5Co4vonw1GFh235-G1Tz1r0NqxNVx+mOxAWCoXCRJJdw45Mp1I59MZIGZrOQ7OonIZPKkfJSAvOCECcHkyAQ7DAyFOYodkuAYlF72Neb4ieTqfTP0z2aLaw2wBlBDlCpASoQBaWtbAecb+Gb2EVkgQ8crWGQamAEECDUaIdVNaNdc2fy6yGQQlEyAxGIANHw1-JFMobOoxNu97w1xZ5NZbGfd-vkK7cR64GI-UloSOM+I+0EzAAmpO06zmoABC8qkNQhQAHLUJmuwLrm9a9v2g7Kiuj4blA94XgeChKCop5bg+l7INeVgnnY56Ps+7r4u+PQQJ+fBAA)

Reply:
Gotcha. Let's use the `export function` format/wrapping I suggested below though